src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,623 @@
+    neighborhood_grants = interf.findOverlappingGrants(
+                            grants_inside, protection_constraint)
+
+    num_unsatisfied_grants = len(neighborhood_grants)
+    
+    # Get number of grants, IAP protection threshold and fairshare per channel 
+    if num_unsatisfied_grants > 0:
+      # algorithm to calculate interference within IAPBW using EIRP obtained for all 
+      # the grants through application of IAP
+
+      # Initialize list of aggregate interference from all the grants 
+      # ( including grants from managing and peer SAS )
+      aggr_interf = [0] * len(channels)
+      # Initialize list of aggregate interference from all the grants 
+      # ( including only grants from managing SAS )
+      asas_interf = [0] * len(channels)
+
+      num_unsatisfied_grants_channel = []
+      iap_threshold_channel = []
+      fairshare_channel = []
+
+      for index, channel in enumerate(channels):
+        # Get protection constraint over 5MHz channel range
+        channel_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=channel[0],
+                               high_frequency=channel[1], entity_type=protection_ent_type)
+
+        # Identify CBSD grants overlapping with frequency range of the protection entity 
+        # in the neighborhood of the protection point and channel
+        channel_neighborhood_grants = interf.findOverlappingGrants(
+                                        neighborhood_grants, channel_constraint)
+
+        num_unsatisfied_grants_channel.append(len(channel_neighborhood_grants))
+        
+        # Set interference quota for protection_point, channel combination
+        # Calculate Roll of Attenuation for ESC Sensor Protection entity
+        if protection_ent_type is interf.ProtectedEntityType.ESC:
+          if channel[0] >= 3650.e6:
+            center_freq = (channel[0] + channel[1]) / 2
+            iap_threshold_channel.append(interf.dbToLinear(interf.linearToDb(threshold) + 
+              (2.5 + ((center_freq - interf.ESC_CH21_CF_HZ) / MHZ))))
+          else:
+            iap_threshold_channel.append(threshold)
+        else:
+          iap_threshold_channel.append(threshold)
+
+        # calculate the fair share per channel based on unsatisfied grants
+        if num_unsatisfied_grants_channel[index] > 0:
+          fairshare_channel.append(float(iap_threshold_channel[index]) / 
+            num_unsatisfied_grants_channel[index] )
+        else:
+          fairshare_channel.append(0)
+
+      # List of grants initialized with max EIRP
+      grants_eirp = [grant.max_eirp for grant in neighborhood_grants]
+      
+      # Initialize list of grants_satisfied to False
+      grants_satisfied = [False] * num_unsatisfied_grants
+      
+      while num_unsatisfied_grants > 0:
+
+        for g_idx, grant in enumerate(neighborhood_grants):
```

Comment:
It seems that the IAP result for a given grant will be different depending on how the "neighborhood_grants" list is ordered (because  fairshare_channel[ch_idx]  is updated within this loop, line 234).

Is it assumed that the "neighborhood_grants"  list is ordered identically by each SAS?

Reply:
The IAP result shall not be dependent of the order of temp lists. Otherwise it is an issue and non deterministic.
Please correct

Reply:
Updated   logic to consider all the grants for all the channels before updating   fairshare